enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -128,154 +101,104 @@ New *pipeline* field:
+  - `openshift.labels.<key>`: Use the string value of an openshift label with key `<key>` (see [](forwarder-tagging.md))
+
+**Note:** If both `structuredIndexName` and `structuredIndexKey` are set,
+the `structuredIndexKey` value is used if it is not empty, `structuredIndexName` otherwise.
 
-* outputDefaults (type map, optional):
-  Each map under outputDefaults MUST have a `type` field.
-  All other fields are used as defaults for any outputs of the same `type` _including_ the `default` output
+**Note:** If the rules above do not give a non-empty index name, the record will be forwarded _unstructured_: with a `message` field and no `structured` field.
+This avoids sending unknown JSON records to shared indices.
+
+Structured indices are created automatically by the managed default store.
+In order to forward to an external Elasticsearch instance, indices must be created in advance.
 
 ### User Stories
 
-#### I want the  default Elasticsearch to index by k8s label on the source pod
+#### I want to forward JSON logs to a remote destination that is not Elasticsearch
+
+This example shows a remote `fluentd` but the same applies to any output other than Elasticsearch:
 
 ```yaml
-outputDefaults:
-- type: elasticsearch
-  indexKey: kubernetes.labels.myIndex
+outputs:
+- name: myFluentd
+  type: fluentdForward
+  url: ...
 
 pipelines:
 - inputRefs: [ application ]
-  outputRefs: default
+  outputRefs: myFluentd
   structured: json
 ```
 
-For example, this log record will go to the index `flintstones`:
+#### I want to forward to default Elasticsearch, using a k8s pod label to determine the index
 
-```json
-{
-  "structured":{
-    "name":"fred",
-    "home":"bedrock",
-  }",
-  "kubernetes":{"labels":{"myIndex": "flintstones", ...}},
-  ...
-}
-```
+For elasticsearch outputs, we must separate logs with different formats into different indices.
+Lets assume that:
 
-Logs from pods with no `myIndex` label will go to the `app` index like any unstructured `message` record.
-
-```json
-{
-  "message":"{\"name\":\"fred\",\"home\":\"bedrock\"}",
-  ... other metadata
-}
-```
-
-#### I want input selector to set the elasticsearch index
-
-Pipelines can set an `openshift` label on records from input selectors.
-This is used the same way as the kubernetes label in the previous example:
+* Applications log in two structured JSON formats called "apache" and "google".
+* User labels pods using those formats with `logFormat=apache` or `logFormat=google`
+* Use the following forwarder configuration:
 
 ```yaml
 outputDefaults:
-- type: elasticsearch
-  indexKey: openshift.labels.myIndex
-
-inputs:
-- name: InputItchy
-  application:
-    namespaces: [ fred, bob ]
-- name: InputScratchy
-  application:
-    namespaces: [ jill, jane ]
-
-pipeline:
-- inputRefs: [ InputItchy ]
-  outputRefs: [ default ]
-  structured: json
-  labels: { myIndex: itchy }
-- inputRefs: [ InputScratchy ]
-  outputRefs: [ default ]
+- elasticsearch:
```

Comment:
Is this redundant? there is type: elasticsearch already.

Reply:
We don't have `type: elasticsearch` any more, you may be seeing that in the deleted part of the diff. The updated document reads:
```yaml
outputDefaults:
- elasticsearch:
    structuredIndexKey: kubernetes.labels.logFormat

pipelines:
- inputRefs: [ application ]
  outputRefs: default
  structured: json
```